network/p2p/connection.go

```diff
@@ -82,32 +87,166 @@ func (conn *connection) send(message *transport.NetworkMessage) error {
 	return nil
 }
 
-// sendMessages (blocking) reads messages from its outMessages channel and sends them to the peer until the channel is closed.
-func (conn connection) sendMessages() {
-	for message := range conn.outMessages {
-		if conn.messenger.Send(message) != nil {
-			log.Logger().Warnf("Unable to send message to peer (peer=%s)", conn.Peer)
+func (conn *managedConnection) exchange(receivedMessages messageQueue) {
+	conn.mux.Lock()
+	out := conn.outMessages
+	in := conn.receiveMessages()
+	conn.mux.Unlock()
+	for {
+		select {
+		case message := <-out:
+			if message == nil {
+				// Connection is closing
+				return
+			}
+			if conn.messenger.Send(message) != nil {
+				log.Logger().Warnf("Unable to send message to peer (peer=%s)", conn.Peer)
+			}
+		case message := <-in:
+			if message == nil {
+				// Connection is closing
+				return
+			}
+			receivedMessages.c <- *message
+		case <-conn.closer:
+			log.Logger().Trace("close() is called")
+			return
 		}
 	}
 }
 
-// receiveMessages (blocking) reads messages from the peer until it disconnects or the network is stopped.
-func (conn *connection) receiveMessages(receivedMsgQueue messageQueue) {
-	for {
-		msg, recvErr := conn.messenger.Recv()
-		if recvErr != nil {
-			if recvErr == io.EOF {
-				log.Logger().Infof("Peer closed connection (peer-id=%s)", conn.ID)
-			} else {
-				log.Logger().Warnf("Peer connection error (peer-id=%s): %v", conn.ID, recvErr)
-			}
-			break
+func (conn *managedConnection) close() {
+	conn.mux.Lock()
+	defer conn.mux.Unlock()
+	if conn.outMessages == nil {
+		// Already closed
+		return
+	}
+	log.Logger().Debugf("Connection is closing (peer-id=%s)", conn.ID)
+
+	// Signal send/receive loop connection is closing
+	if len(conn.closer) == 0 {
+		conn.closer <- struct{}{}
+	}
+	// Close our client connection (not applicable if we're the server side of the connection)
+	if conn.grpcConn != nil {
+		if err := conn.grpcConn.Close(); err != nil {
+			log.Logger().Warnf("Unable to close client connection (peer=%s): %v", conn.Peer, err)
 		}
-		log.Logger().Tracef("Received message from peer (peer-id=%s): %s", conn.ID, msg.String())
-		receivedMsgQueue.c <- PeerMessage{
-			Peer:    conn.ID,
-			Message: msg,
+		conn.grpcConn = nil
+	}
+	// Close in/out channels
+	close(conn.outMessages)
+	conn.outMessages = nil
+}
+
+func (conn *managedConnection) receiveMessages() chan *PeerMessage {
+	peerID := conn.ID
+	messenger := conn.messenger
+	result := make(chan *PeerMessage, 10)
+	go func() {
+		for {
+			msg, recvErr := messenger.Recv()
```

Comment:
What happens to this call when we close the connection?

Reply:
This is a bit complex and depends on the side of the connection:

- client side `Recv()` is cancelled by closing the `ClientStream` (status = cancelled)
- server side `Recv()` is cancelled by letting the gRPC service handler (implementation of `Connect()`) exit, which closes the underlying gRPC channel, which in turn cancels `Recv()` (status = EOF)

I documented this behaviour a bit better.

Reply:
👍